### PR TITLE
Add early stopping to trainer_tf.py

### LINKED ISSUE #4894: 🚀 Add early stopping to the trainer
# 🚀 Feature request

The trainer ([pt](https://github.com/huggingface/transformers/blob/master/src/transformers/trainer.py), [tf](https://github.com/huggingface/transformers/blob/master/src/transformers/trainer_tf.py)) is an easy access point for users who rather not spend too much time building their own trainer class but prefer an out-of-the-box solution. Even though `transformers` was never meant to be a fully fletched training library, it might please users to add an additional feature: early stopping. 

## Motivation

Early stopping ensures that the trainer does not needlessly keep training when the loss does not improve. This saves time, money, and let's not forget the trees. 😉 Performance-wise this should not lead to different results.

## Your contribution

At the moment I cannot work on this, but here are my thoughts:

- a training argument should be added ([pt](https://github.com/huggingface/transformers/blob/master/src/transformers/training_args.py), [tf](https://github.com/huggingface/transformers/blob/master/src/transformers/trainer_tf.py)). This would only work when `evaluate_during_training` is enabled.
- for PyTorch: at every evaluation step, an early stopper (can be a separate class even) checks if the loss has improved in the last n steps. Potentially with a minimal threshold that the loss should have improved. If not, the trainer should stop
- for Tensorflow: I don't have experience with TF myself, but I assume one could use [`tf.keras.callbacks.EarlyStopping`](https://www.tensorflow.org/api_docs/python/tf/keras/callbacks/EarlyStopping).

### PR DESCRIPTION
**Pull request by KMFODA**
## Summary
This PR adds the early stopping feature to trainer_tf.py.

## Related Issues
Alongside #4186 should close #4894.

## Who can review?

@sgugger & @BramVanroy.

### CODE REVIEW

**Review by sgugger — COMMENTED** [review]
We can definitely start with this but ideally, I'd like `TFTrainer` to support early stopping with a metric, not just the validation loss. Introducing the arguments `metric_for_best_model` and `greater_is_better` like in the PyTorch Trainer is on the roadmap for TFTrainer, and they could be used for this too!

Up to you if you want to add this to your PR or if we just merge it as is for now and I or @jplu will add the rest of the behavior later on.

**Comment by jplu** [comment]
Thanks @KMFODA!

I'm not really in favor to do this manually as there are already a Keras callbacks taking care of this https://www.tensorflow.org/api_docs/python/tf/keras/callbacks/EarlyStopping that can monitor multiple values on which to base his early stop.

@KMFODA if you want to add the feature to add Keras callbacks handling, it will be more than welcome :) As it is part of the features you plan to add in a near future.

**Comment by BramVanroy** [comment]
Seeing that I have close to no experience with TF, I won't be able to review this.

**Comment by KMFODA** [comment]
Not a problem. @jplu I agree prefer using Keras callbacks I just only have experience using it with Keras’s model.fit function. I’ll think and experiment with how to use it in a custom built TF model. Hopefully if successful, it should be fairly simple then to add early stopping based on custom metrics rather than just validation loss.

**Comment by KMFODA** [comment]
I've just pushed the latest changes to trainer_tf.py that will use Keras's callbacks for early stopping rather than the manual solution I had initially submitted. Setting the callback in the `TFTrainer` function using a command such as this:

`callbacks = [EarlyStopping(monitor='loss', patience=1, verbose=1)]
`

will monitor the training loss and stop the model at the first epoch which fails to improve on the best training loss metric. Using the monitor variable we can also select metrics, other than the training loss, to carry out the early stopping on.

**Comment by KMFODA** [comment]
Hi all, anything more I can do to help get this merged?

**Comment by BramVanroy** [comment]
Bump @LysandreJik @jplu

**Comment by sgugger** [comment]
We're moving away from the TFTrainer to fully integrate with Keras, so we won't add new functionality to the TFTrainer.

**Comment by BramVanroy** [comment]
> We're moving away from the TFTrainer to fully integrate with Keras, so we won't add new functionality to the TFTrainer.

Alright, sounds good, This can be closed then?

**Comment by sgugger** [comment]
I'll let the original author close it :-)
